flink-table-store-core/src/main/java/org/apache/flink/table/store/file/operation/AbstractFileStoreScan.java

```diff
@@ -272,4 +272,15 @@ private boolean filterByBucket(ManifestEntry entry) {
     private List<ManifestEntry> readManifestFileMeta(ManifestFileMeta manifest) {
         return manifestFileFactory.create().read(manifest.fileName());
     }
+
+    private List<ManifestFileMeta> readIncremental(Long snapshotId) {
```

Comment:
Why do this change?
I think we can add a validation here `snapshot.commitKind() == Snapshot.CommitKind.APPEND`, but why a `while`?

Reply:
> Why do this change?
I think we can add a validation here `snapshot.commitKind() == Snapshot.CommitKind.APPEND`, but why a while?

To try best to find the latest `APPEND`. Because the commit order is new files first, then compact before and after. So if auto-compaction continuously happened during streaming write, then the snapshot will be like

- txn1: `APPEND`(snapshot-1), `COMPACT`(snapshot-2)
- txn2: `APPEND`(snapshot-3), `COMPACT`(snapshot-4)
- ...

If we want to try our best to support incremental reading, we should find the latest append snapshot instead of throwing exceptions?

Reply:
I think it is better to throw exception.
From an API perspective, this semantics you describe is too complex and we don't rely on it.